File: DONGUK/프로그래머스_WEEK_10/방문 길이.py
def solution(dirs):
    answer = 0
    visited = []
    
    x, y = 0, 0
    for c in dirs:
        p_x, p_y = x, y
        if c == 'U':
            if y < 5:
                y += 1
            else:
                continue
                
        elif c == 'D':
            if y > -5:
                y -= 1
            else:
                continue
        
        elif c == 'L':
            if x > -5:
                x -= 1
            else:
                continue
            
        else:
            if x < 5:
                x += 1
            else:
                continue
        
        if (p_x, p_y, x, y) not in visited and (x, y, p_x, p_y) not in visited:
            visited.append((p_x, p_y, x, y))
            answer += 1
            
    return answer

File: JINSEOP/프로그래머스_WEEK10/방문 길이.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <set>
using namespace std;

set<vector<pair<int, int>>> classifier;
void move_character(string dirs) {
    pair<int, int> coordinate = { 0, 0 };

    for (int i = 0; i < dirs.size(); i++) {
        vector<pair<int, int>> temp;
        switch (dirs[i])
        {
        case 'U':
            if (coordinate.second < 5) {
                temp.push_back(coordinate);
                coordinate.second++;
                temp.push_back(coordinate);
                sort(temp.begin(), temp.end());
                classifier.insert(temp);
            }
            break;
        case 'D':
            if (coordinate.second > -5) {
                temp.push_back(coordinate);
                coordinate.second--;
                temp.push_back(coordinate);
                sort(temp.begin(), temp.end());
                classifier.insert(temp);
            }
            break;
        case 'R':
            if (coordinate.first < 5) {
                temp.push_back(coordinate);
                coordinate.first++;
                temp.push_back(coordinate);
                sort(temp.begin(), temp.end());
                classifier.insert(temp);
            }
            break;
        case 'L':
            if (coordinate.first > -5) {
                temp.push_back(coordinate);
                coordinate.first--;
                temp.push_back(coordinate);
                sort(temp.begin(), temp.end());
                classifier.insert(temp);
            }
            break;
        }
    }
}

int solution(string dirs) {
    int answer = 0;
    
    move_character(dirs);
    answer = classifier.size();

    return answer;
}
